chapter_4/exercise_4_04/stack.c

```diff
@@ -158,10 +158,10 @@ void swap(void)
 
 void clear(void)
 {
-  do
+  while (sp > 0)
   {
-    stack[sp] = 0.0;
-  } while (sp--);
+    stack[--sp] = 0.0;
+  }
```

Comment:
@sadeem-albir Is this code required to change? What's the bug here? Is it that pre decrement that useful?

Reply:
@ohkimur If my assumption is correct, the sp works the same as buf and bufp. whenever you add something to the stack, you do stack[sp++], meaning the current index (zero) will be assigned then sp will be the next unassigned index. So by using that other while-loop instead, you ensure you move to the previous assigned sp until you reach zero. Also, in the do-loop, even if sp was zero, it will be decremented, undesirably becoming -1.

Reply:
@ohkimur Not necessarily a negative impact. As you see in the pop function, you return stack[--sp] so sp is meant to be empty at whatever current index it is lying at for the moment, and when we need to use a value from the stack, we simply call the previous index of sp, so if the stack was [1, 2, 3, 0] and sp is 3, calling pop will return the value of 3 at index 2. The change I have proposed ensures consistency with the way sp changes in the other stack functions, but if you believe staying at the previous change is better for the program, then the change doesn't have to be merged.